cub/cub/device/dispatch/dispatch_segmented_scan.cuh

```diff
@@ -0,0 +1,374 @@
+    BeginOffsetIteratorOutputT,
+    OffsetT,
+    ScanOpT,
+    InitValueT,
+    AccumT,
+    EnforceInclusive>,
+  typename KernelLauncherFactory = CUB_DETAIL_DEFAULT_KERNEL_LAUNCHER_FACTORY>
+struct DispatchSegmentedScan
+{
+  static_assert(::cuda::std::is_integral_v<OffsetT> && sizeof(OffsetT) >= 4 && sizeof(OffsetT) <= 8,
+                "DispatchSegmentedScan only supports integral offset types of 4- or 8-bytes");
+
+  /// Device-accessible allocation of temporary storage.  When nullptr, the
+  /// required allocation size is written to \p temp_storage_bytes and no work
+  /// is done.
+  void* d_temp_storage;
+
+  /// Reference to size in bytes of \p d_temp_storage allocation
+  size_t& temp_storage_bytes;
+
+  /// Iterator to the input sequence of data items
+  InputIteratorT d_in;
+
+  /// Iterator to the output sequence of data items
+  OutputIteratorT d_out;
+
+  /// The number of segments that comprise the segmented scan data
+  ::cuda::std::int64_t num_segments;
+
+  BeginOffsetIteratorInputT d_input_begin_offsets;
+
+  EndOffsetIteratorInputT d_input_end_offsets;
+
+  BeginOffsetIteratorOutputT d_output_begin_offsets;
+
+  /// Binary scan functor
+  ScanOpT scan_op;
+
+  /// Initial value to seed the exclusive scan
+  InitValueT init_value;
+
+  /// CUDA stream to launch kernels within. Default is stream<sub>0</sub>.
+  cudaStream_t stream;
+
+  int ptx_version;
+
+  KernelSource kernel_source;
+
+  KernelLauncherFactory launcher_factory;
+
+  // Constructor
+  CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE DispatchSegmentedScan(
+    void* d_temp_storage,
+    size_t& temp_storage_bytes,
+    InputIteratorT d_in,
+    OutputIteratorT d_out,
+    ::cuda::std::int64_t num_segments,
+    BeginOffsetIteratorInputT d_in_begin_offsets,
+    EndOffsetIteratorInputT d_in_end_offsets,
+    BeginOffsetIteratorOutputT d_out_begin_offsets,
+    ScanOpT scan_op,
+    InitValueT init_value,
+    cudaStream_t stream,
+    int ptx_version,
+    KernelSource kernel_source             = {},
+    KernelLauncherFactory launcher_factory = {})
+      : d_temp_storage(d_temp_storage)
+      , temp_storage_bytes(temp_storage_bytes)
+      , d_in(d_in)
+      , d_out(d_out)
+      , num_segments(num_segments)
+      , d_input_begin_offsets(d_in_begin_offsets)
+      , d_input_end_offsets(d_in_end_offsets)
+      , d_output_begin_offsets(d_out_begin_offsets)
+      , scan_op(scan_op)
+      , init_value(init_value)
+      , stream(stream)
+      , ptx_version(ptx_version)
+      , kernel_source(kernel_source)
+      , launcher_factory(launcher_factory)
+  {}
+
+  template <typename ActivePolicyT, typename SegmentedScanKernelT>
+  CUB_RUNTIME_FUNCTION _CCCL_HOST _CCCL_FORCEINLINE cudaError_t
+  InvokePasses(SegmentedScanKernelT segmented_scan_kernel, ActivePolicyT policy = {})
+  {
+    // `LOAD_LDG` makes in-place execution UB and doesn't lead to better
+    // performance.
+    policy.CheckLoadModifier();
+
+    cudaError error = cudaSuccess;
+    do
+    {
```

Comment:
Strong suggestion: Please avoid the `do` loop here and just return in case of errors instead of breaking.